pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_reconcile.go

```diff
@@ -0,0 +1,146 @@
+	// For now the unfiltered list is added.
+
+	// Get APIExport
+	apiExportPath := logicalcluster.NewPath(apiExportEndpointSlice.Spec.APIExport.Path)
+	if apiExportPath.Empty() {
+		apiExportPath = logicalcluster.From(apiExportEndpointSlice).Path()
+	}
+	apiExport, err := c.getAPIExport(logicalcluster.From(apiExportEndpointSlice), apiExportPath, apiExportEndpointSlice.Spec.APIExport.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Don't keep the endpoints if the APIExport has been deleted
+			apiExportEndpointSlice.Status.APIExportEndpoints = nil
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.APIExportNotFoundReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"APIExport %s|%s not found",
+				apiExportPath,
+				apiExportEndpointSlice.Spec.APIExport.Name,
+			)
+			return nil
+		} else {
+			conditions.MarkFalse(
+				apiExportEndpointSlice,
+				apisv1alpha1.APIExportValid,
+				apisv1alpha1.InternalErrorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"Error getting APIExport %s|%s",
+				apiExportPath,
+				apiExportEndpointSlice.Spec.APIExport.Name,
+			)
+			return err
+		}
+	}
+	conditions.MarkTrue(apiExportEndpointSlice, apisv1alpha1.APIExportValid)
+
+	if err = c.updateEndpoints(ctx, apiExportEndpointSlice, apiExport); err != nil {
+		conditions.MarkFalse(
+			apiExportEndpointSlice,
+			apisv1alpha1.APIExportEndpointSliceURLsReady,
+			apisv1alpha1.ErrorGeneratingURLsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+		return err
+	}
+	conditions.MarkTrue(apiExportEndpointSlice, apisv1alpha1.APIExportEndpointSliceURLsReady)
+
+	return nil
+}
+
+func (c *controller) updateEndpoints(ctx context.Context,
+	apiExportEndpointSlice *apisv1alpha1.APIExportEndpointSlice,
+	apiExport *apisv1alpha1.APIExport) error {
+	logger := klog.FromContext(ctx)
+	shards, err := c.listShards()
+	if err != nil {
+		// Don't update the endpoint list if the cache server could not be reached
+		// but provide a minimal, local list if none was populated
+		if apiExportEndpointSlice.Status.APIExportEndpoints != nil || len(shards) == 0 {
+			return fmt.Errorf("error listing Shards: %w", err)
+		}
+	}
+
+	desiredURLs := sets.NewString()
+	for _, shard := range shards {
+		logger = logging.WithObject(logger, shard)
+		if shard.Spec.VirtualWorkspaceURL == "" {
+			continue
+		}
+
+		u, err := url.Parse(shard.Spec.VirtualWorkspaceURL)
+		if err != nil {
+			// Should never happen
+			logger.Error(
+				err, "error parsing Shard.Spec.VirtualWorkspaceURL",
+				"VirtualWorkspaceURL", shard.Spec.VirtualWorkspaceURL,
+			)
+
+			continue
+		}
+
+		u.Path = path.Join(
+			u.Path,
+			virtualworkspacesoptions.DefaultRootPathPrefix,
+			apiexportbuilder.VirtualWorkspaceName,
+			logicalcluster.From(apiExport).String(),
+			apiExport.Name,
+		)
+
+		desiredURLs.Insert(u.String())
+	}
+
+	apiExportEndpointSlice.Status.APIExportEndpoints = nil
```

Comment:
If the list of shards is empty (the cache server is unavailable) this will wipe previously set endpoints.

I think in such a case we should fallback on `c.ShardVirtualWorkspaceURL()` but only for new `apiExportEndpointSlice` - so that we won't erase previous work.

Reply:
I think this is incorrect, we could improve by waiting for `waitForOptionalSync` before starting your controller.

Reply:
done